### ISSUE TITLE
romio: hang with deferred open feature

### ISSUE THREAD

**Issue by jhendersonHDF** [pmodels/mpich#6408]
Hi All, I'm observing a hang in ROMIO when using netCDF/HDF5 with MPICH while enabling the deferred open feature with the "romio_no_indep_rw" MPI hint. I have tested with the latest develop (commit 30f3f0e) and see the same issue. Please refer to https://github.com/HDFGroup/hdf5/issues/2434 for a netCDF reproducer; I can probably create a pure HDF5 example if that would be more helpful.

When I run the example in that issue with two ranks, I observe that one rank gets stuck in an `MPI_Bcast` call in `ADIOI_Info_check_and_install_enabled`, while the other rank typically gets stuck in the `MPI_Barrier` call [here](https://github.com/pmodels/mpich/blob/main/src/mpi/romio/mpi-io/set_view.c#L169). From what I can tell, the following appears to be happening:

During a collective file open with the deferred open feature enabled, the "striping_unit" hint is set on non-aggregator ranks [here](https://github.com/pmodels/mpich/blob/main/src/mpi/romio/adio/common/ad_opencoll.c#L127) and then they return right away. Later on, during the call to `nc_close` in that example program, a call to `MPI_File_set_view` is called. During that, both ranks proceed to call [ADIO_Set_view](https://github.com/pmodels/mpich/blob/main/src/mpi/romio/mpi-io/set_view.c#L141) -> [ADIO_SetInfo](https://github.com/pmodels/mpich/blob/main/src/mpi/romio/adio/common/ad_set_view.c#L69) -> [ADIOI_GEN_SetInfo](https://github.com/pmodels/mpich/blob/main/src/mpi/romio/adio/common/ad_hints.c#L11). In `ADIOI_GEN_SetInfo`, both ranks go through the process of checking the MPI info passed in from the user, until getting [here](https://github.com/pmodels/mpich/blob/main/src/mpi/romio/adio/common/ad_hints.c#L260-L266).

From what I can see, both ranks have all of the different MPI hints being checked for set on their respective MPI info objects, except that the "striping_unit" hint is only set for one rank. When both ranks get to the call to `ADIOI_Info_check_and_install_int` for that hint, one rank skips most of the function because that hint isn't set, while the other rank hits the call to `MPI_Bcast` within that function. The rank that didn't have the hint set proceeds to the following `ADIOI_Info_check_and_install_enabled` call for the "romio_synchronized_flush" hint and, since it has that hint set, calls MPI_Bcast within that function. Since both the `MPI_Bcast` calls are broadcasting a value of `MPI_INT`, they match each other. Then, the rank that's ahead proceeds out of `ADIOI_GEN_SetInfo` and into the previously mentioned `MPI_Barrier` call in `MPI_File_set_view`, while the trailing rank proceeds on to `ADIOI_Info_check_and_install_enabled` for the "romio_synchronized_flush" hint and both ranks hang.

I suspect that this is probably related to [this comment](https://github.com/pmodels/mpich/blob/main/src/mpi/romio/adio/common/ad_hints.c#L124-L125) where the info object doesn't get initialized with a "striping_unit" hint since there isn't a good default to choose really.

I've attached a patch that fixes the hang (for some reason GitHub won't allow me to upload a .patch file..), but I understand it's not a real solution for the issue.

[striping_unit_hang.txt](https://github.com/pmodels/mpich/files/10762189/striping_unit_hang.txt)

**Comment by hzhou** [comment]
@jhendersonHDF Just for trouble shoot, can you try configure mpich with ch3 `--with-device=ch3` and see if the issue persist?

**Comment by jhendersonHDF** [comment]
Hi @hzhou, yes it looks like the hang still occurs even when MPICH is configured with `--with-device=ch3`. The test program appears to not hang with the HDF5 1.13.1 release, but investigating that seems to show that the 1.13.1 release probably wasn't correctly passing down the application's MPI Info object.

**Comment by jhendersonHDF** [comment]
Happy to be of help if that patch looks almost serviceable! Is the checking of those hints on a performance critical path? If not, would it make any sense to have something like an "is this flag set" matrix and then add something like an `MPI_Allreduce` for that matrix so that processors can get together and determine if any of the ranks have a particular hint set? That's probably a sledgehammer solution, but it might be one way to avoid having a processor accidentally skip one of those MPI_Bcast calls in the future.

**Comment by roblatham00** [comment]
the standard helps us here by requiring hints to be set and the same on all processes.  the "check if all the same" is not strictly required and could be removed, but I suspect if someone did different values for some of ROMIO's hints that might cause some odd hard to track down behavior.

**Comment by jhendersonHDF** [comment]
Ah ok. I never specifically looked into what the standard mentions about MPI hints, but makes sense to me. Thanks for looking into this issue though; I can definitely understand the complications here behind verifying good behavior from the user/application.

**Comment by roblatham00** [comment]
I'm looking at your proposed fix a bit more closely today...  I believe you when you tell me it fixes the problem but I still have a few questions:
- why just the striping_unit hint?  striping_factor is set the same way and yet not part of the workaround
- what is HDF5 doing to trigger this?   I wrote a few MPI-IO test cases to try to trigger this bug and was not able to.

**Comment by jhendersonHDF** [comment]
Hi @roblatham00,

> why just the striping_unit hint? striping_factor is set the same way and yet not part of the workaround

I'm not exactly sure whether `striping_unit` is the only problematic hint here, but appears to be one of them. Based on the code in https://github.com/pmodels/mpich/blob/main/src/mpi/romio/adio/common/ad_hints.c#L59-L151, it also looks like no default value is set for the `striping_factor` or `romio_lustre_start_iodevice` hints that get set on aggregators, but the code in the immediately following block https://github.com/pmodels/mpich/blob/main/src/mpi/romio/adio/common/ad_hints.c#L154-L267 is only checking the `striping_unit` hint with `ADIOI_Info_check_and_install_int`. I don't see a similar check for the `striping_factor` or `romio_lustre_start_iodevice` hints, which is probably why the `striping_unit` hint is the problematic one, though other checks for the additional hints might occur elsewhere in the MPICH code?

> what is HDF5 doing to trigger this? I wrote a few MPI-IO test cases to try to trigger this bug and was not able to.

I'm not sure here whether HDF5 is doing something specific, but it seemed like just setting the "romio_no_indep_rw" hint was enough to cause the issue when using HDF5 with netCDF and MPICH/romio. I can do my best to create a minimal HDF5 example; I just haven't tested yet to see if that exhibits the same issue. I certainly haven't tried yet with just MPI I/O.

**Comment by roblatham00** [comment]
I have a personal affection for deferred open -- it was my first ROMIO feature all those years ago.  but every time people poke at it aggressively we find little bugs like this.  So I want to add more test cases to ROMIO and MPICH... but  this test case was not sufficient to trigger the hang:
```
int main(int argc, char* argv[])
{
    int my_rank;
    MPI_Info info;
    MPI_File fh;

    MPI_Init(&argc, &argv);

    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

    MPI_Info_create(&info);
    MPI_Info_set(info, "romio_no_indep_rw", "true");

    CHECK(MPI_File_open(MPI_COMM_WORLD, argv[1], MPI_MODE_CREATE|MPI_MODE_WRONLY, info, &fh));
    CHECK(MPI_File_set_view(fh, 0, MPI_INT, MPI_INT, "native", info));

    CHECK(MPI_File_close(&fh));
    CHECK(MPI_Info_free(&info));

    MPI_Finalize();

    return 0;
}
```

**Comment by roblatham00** [comment]
You're right that there are three hints being set and we might end up hanging like this again one day in the future if we don't provide defaults for all three.    Maybe I should set these hints to "UNDEFINED" -- hints are strings after all.  then when ROMIO parses them it can special case "UNDEFINED" before calling  `atoii`

**Comment by jhendersonHDF** [comment]
> I have a personal affection for deferred open -- it was my first ROMIO feature all those years ago.

I certainly know the feeling!

> Maybe I should set these hints to "UNDEFINED" -- hints are strings after all. then when ROMIO parses them it can special case "UNDEFINED" before calling atoii

I think setting defaults is reasonable and should fix the hang issue, but I guess it's also not clear at this point if it's really a bug in MPICH or if it's HDF5 misbehaving as middleware. Let me try to create an HDF5 only example and if that still hangs, then we can work down towards an MPI I/O one or see what HDF5 is doing wrong.

**Comment by jhendersonHDF** [comment]
@roblatham00 Just a bit more information. I traced every MPI call on the way down to the hang in the program above for 2 MPI ranks and everything seems to be the same except that after the call to MPI_File_open, HDF5 calls MPI_File_get_info on the file and then copies the Info's hints back into the Info object used for the MPI_File_open_call, effectively overwriting what was there before. The relevant code and comments can be found at: https://github.com/HDFGroup/hdf5/blob/develop/src/H5FDmpio.c#L906-L956. This was added in commit https://github.com/HDFGroup/hdf5/commit/aaa6614c489d68fd996bdc5c393488548f7c15a6.

The issue seems to be that the aggregator ranks are then the only ones that end up with the `striping_unit`, `striping_factor` and `romio_lustre_start_iodevice` hints on the MPI Info object for their internal file structure. Then, later calls to MPI_File_set_view are only only going to have `striping_unit` set for aggregator ranks. I modified your example to effectively do what HDF5 is doing now. I don't know if this should be considered proper behavior or not.

```C
#include <mpi.h>

int main(int argc, char* argv[])
{
    int my_rank;
    MPI_Info info;
    MPI_Info info_used;
    MPI_File fh;

    MPI_Init(&argc, &argv);

    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);

    MPI_Info_create(&info);
    MPI_Info_set(info, "romio_no_indep_rw", "true");

    MPI_File_open(MPI_COMM_WORLD, argv[1], MPI_MODE_CREATE|MPI_MODE_WRONLY, info, &fh);
    MPI_File_get_info(fh, &info_used);

    if (info_used != MPI_INFO_NULL) {
        int i, nkeys;

        MPI_Info_get_nkeys(info_used, &nkeys);

        for (i = 0; i < nkeys; i++) {
            char key[MPI_MAX_INFO_KEY], value[MPI_MAX_INFO_VAL];
            int  valuelen, flag;

            MPI_Info_get_nthkey(info_used, i, key);
            MPI_Info_get_valuelen(info_used, key, &valuelen, &flag);
            MPI_Info_get(info_used, key, valuelen + 1, value, &flag);

            MPI_Info_set(info, key, value);
        }

        MPI_Info_free(&info_used);
    }

    MPI_File_set_view(fh, 0, MPI_INT, MPI_INT, "native", info);

    MPI_File_close(&fh);
    MPI_Info_free(&info);

    MPI_Finalize();

    return 0;
}
```

**Comment by roblatham00** [comment]
Ah, well that explains why we were in the "process user-provided hints" section further than I expected. 

this kind of hint processing looks perfectly fine to me.    Can I add this as a test case to ROMIO and MPICH?

**Comment by jhendersonHDF** [comment]
Absolutely. Whatever helps with testing MPICH.

**Comment by dqwu** [comment]
@roblatham00 FYI, this issue is still reproducible with latest MPICH release 4.1.2.
Will it be fixed by PR #6494?
